### PR TITLE
Separate layers - Application + Project

### PR DESCRIPTION
**Pull request by feloy**
**What type of PR is this?**

/kind cleanup

**What does this PR do / why we need it**:

**Which issue(s) this PR fixes**:

Part of #5247 

**PR acceptance criteria**:

- [ ] Unit test 

- [ ] Integration test 

- [ ] Documentation 

- [ ] I have read the [test guidelines](https://github.com/redhat-developer/odo/blob/master/docs/dev/test-architecture.adoc)

**How to test changes / Special notes to the reviewer**:

### CODE REVIEW

**Comment by feloy** [comment]
Kubernetes Tests finished successfully.
View logs: [TXT](https://s3.eu-de.cloud-object-storage.appdomain.cloud/odo-tests-openshift-logs/pr-5293-kubernetes-tests-165.txt) [HTML](https://s3.eu-de.cloud-object-storage.appdomain.cloud/odo-tests-openshift-logs/pr-5293-kubernetes-tests-165.html)

**Comment by feloy** [comment]
Unit Tests finished successfully.
View logs: [TXT](https://s3.eu-de.cloud-object-storage.appdomain.cloud/odo-tests-openshift-logs/pr-5293-unit-tests-178.txt) [HTML](https://s3.eu-de.cloud-object-storage.appdomain.cloud/odo-tests-openshift-logs/pr-5293-unit-tests-178.html)

**Comment by feloy** [comment]
OpenShift Tests finished successfully.
View logs: [TXT](https://s3.eu-de.cloud-object-storage.appdomain.cloud/odo-tests-openshift-logs/pr-5293-openshift-tests-194.txt) [HTML](https://s3.eu-de.cloud-object-storage.appdomain.cloud/odo-tests-openshift-logs/pr-5293-openshift-tests-194.html)

**Comment by feloy** [comment]
Kubernetes Tests finished successfully.
View logs: [TXT](https://s3.eu-de.cloud-object-storage.appdomain.cloud/odo-tests-openshift-logs/pr-5293-kubernetes-tests-171.txt) [HTML](https://s3.eu-de.cloud-object-storage.appdomain.cloud/odo-tests-openshift-logs/pr-5293-kubernetes-tests-171.html)

**Comment by feloy** [comment]
Unit Tests finished successfully.
View logs: [TXT](https://s3.eu-de.cloud-object-storage.appdomain.cloud/odo-tests-openshift-logs/pr-5293-unit-tests-184.txt) [HTML](https://s3.eu-de.cloud-object-storage.appdomain.cloud/odo-tests-openshift-logs/pr-5293-unit-tests-184.html)

**Comment by feloy** [comment]
OpenShift Tests finished successfully.
View logs: [TXT](https://s3.eu-de.cloud-object-storage.appdomain.cloud/odo-tests-openshift-logs/pr-5293-openshift-tests-200.txt) [HTML](https://s3.eu-de.cloud-object-storage.appdomain.cloud/odo-tests-openshift-logs/pr-5293-openshift-tests-200.html)

**Comment by feloy** [comment]
Unit Tests finished with errors.
View logs: [TXT](https://s3.eu-de.cloud-object-storage.appdomain.cloud/odo-tests-openshift-logs/pr-5293-unit-tests-190.txt) [HTML](https://s3.eu-de.cloud-object-storage.appdomain.cloud/odo-tests-openshift-logs/pr-5293-unit-tests-190.html)

**Comment by feloy** [comment]
Kubernetes Tests finished successfully.
View logs: [TXT](https://s3.eu-de.cloud-object-storage.appdomain.cloud/odo-tests-openshift-logs/pr-5293-kubernetes-tests-177.txt) [HTML](https://s3.eu-de.cloud-object-storage.appdomain.cloud/odo-tests-openshift-logs/pr-5293-kubernetes-tests-177.html)

**Comment by feloy** [comment]
OpenShift Tests finished successfully.
View logs: [TXT](https://s3.eu-de.cloud-object-storage.appdomain.cloud/odo-tests-openshift-logs/pr-5293-openshift-tests-206.txt) [HTML](https://s3.eu-de.cloud-object-storage.appdomain.cloud/odo-tests-openshift-logs/pr-5293-openshift-tests-206.html)

**Comment by feloy** [comment]
Unit Tests finished successfully.
View logs: [TXT](https://s3.eu-de.cloud-object-storage.appdomain.cloud/odo-tests-openshift-logs/pr-5293-unit-tests-200.txt) [HTML](https://s3.eu-de.cloud-object-storage.appdomain.cloud/odo-tests-openshift-logs/pr-5293-unit-tests-200.html)

**Comment by feloy** [comment]
OpenShift Tests finished with errors.
View logs: [TXT](https://s3.eu-de.cloud-object-storage.appdomain.cloud/odo-tests-openshift-logs/pr-5293-openshift-tests-216.txt) [HTML](https://s3.eu-de.cloud-object-storage.appdomain.cloud/odo-tests-openshift-logs/pr-5293-openshift-tests-216.html)

**Comment by feloy** [comment]
Kubernetes Tests finished with errors.
View logs: [TXT](https://s3.eu-de.cloud-object-storage.appdomain.cloud/odo-tests-openshift-logs/pr-5293-kubernetes-tests-187.txt) [HTML](https://s3.eu-de.cloud-object-storage.appdomain.cloud/odo-tests-openshift-logs/pr-5293-kubernetes-tests-187.html)

**Comment by kadel** [comment]
This is great :+1: I like the approach with `Client` interfaces

/approve

File: pkg/project/kubernetes.go [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+		isActive := project == currentProject
+		projects[i] = NewProject(project, isActive)
+	}
+
+	return NewProjectList(projects), nil
+}
+
+// Exists checks whether a project with the name `projectName` exists and returns an error if any
+func (o kubernetesClient) Exists(projectName string) (bool, error) {
+	projectSupport, err := o.client.IsProjectSupported()
```
Comment (dharmit) [inline pkg/project/kubernetes.go:124]: Why not have the check for empty value of `projectName` here? We have it in `Create` and `Delete` methods. 🤔

**Review by dharmit — CHANGES_REQUESTED** [review]
I understand that at times I have requested/suggested changes in the piece of code that you have only copied over. I have done that since we are sort of refactoring the bits here. If you think we should handle them separately, we can discuss.

File: pkg/application/kubernetes.go [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+
+type kubernetesClient struct {
+	client kclient.ClientInterface
+}
+
+func NewClient(client kclient.ClientInterface) Client {
+	return kubernetesClient{
+		client: client,
+	}
+}
```
[the remark this answers is not in the document]
Comment (feloy) [reply pkg/application/kubernetes.go:23]: We need to have a type, to make it implement the Client interface

File: pkg/application/kubernetes.go [hunk 3]
```diff
@@ -0,0 +1,111 @@
+package application
```
[the remark this answers is not in the document]
Comment (feloy) [reply pkg/application/kubernetes.go:1]: Not particularly, this is to separate the implementation from the definition of the interface, and to make something similar to `url` and   `storage`.

File: pkg/application/kubernetes.go [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+}
+
+func NewClient(client kclient.ClientInterface) Client {
+	return kubernetesClient{
+		client: client,
+	}
+}
+
+// List all applications names in current project by looking at `app` labels in deployments
+func (o kubernetesClient) List() ([]string, error) {
```
[the remark this answers is not in the document]
Comment (feloy) [reply pkg/application/kubernetes.go:26]: That would an interesting subject for another refactoring

File: pkg/application/types.go [hunk 5]
```diff
@@ -4,6 +4,12 @@ import (
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 )
 
+const (
+	appAPIVersion = "odo.dev/v1alpha1"
```
[the remark this answers is not in the document]
Comment (feloy) [reply pkg/application/types.go:8]: Using machineoutput.APIVersion, thanks

File: pkg/odo/cli/application/list.go [hunk 6]
```diff
@@ -54,52 +62,54 @@ func (o *ListOptions) Validate() (err error) {
 
 // Run contains the logic for the odo command
 func (o *ListOptions) Run() (err error) {
-	apps, err := application.List(o.KClient)
+	apps, err := o.appClient.List()
 	if err != nil {
 		return fmt.Errorf("unable to get list of applications: %v", err)
 	}
```
[the remark this answers is not in the document]
Comment (feloy) [reply pkg/odo/cli/application/list.go:68]: I'm not sure. The CLI layer should try to help the user understand the error, here by giving some more context

File: pkg/project/kubernetes.go [hunk 7]
```diff
@@ -0,0 +1,142 @@
...
+		}
+	}
+	return nil
+}
+
+// Delete deletes the project (the `project` resource if supported, or directly the `namespace`)
+// with the name projectName and returns an error if any
+func (o kubernetesClient) Delete(projectName string, wait bool) error {
+	if projectName == "" {
+		return errors.Errorf("no project name given")
```
[the remark this answers is not in the document]
Comment (feloy) [reply pkg/project/kubernetes.go:68]: Instead, I made the same change as in List to have to return the error only once

File: pkg/project/kubernetes.go (hunk 1 above)
Comment (feloy) [reply pkg/project/kubernetes.go:124]: I think that this specific case is detected by `GetProject`/`GetNamespace`

File: pkg/application/kubernetes.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (dharmit) [reply pkg/application/kubernetes.go:27]: Sure. Let's discuss this alongside the v3 plans.

File: pkg/odo/cli/application/list.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (dharmit) [reply pkg/odo/cli/application/list.go:68]: Makes sense.

**Pull request merged**